packages/web-wallet/src/services/SnapService.ts

```diff
@@ -0,0 +1,193 @@
+import { HATHOR_API_URLS, NETWORKS, TOKEN_IDS, DEFAULT_NETWORK } from '../constants';
+import type { WalletBalance } from '../types/wallet';
+import { createLogger } from '../utils/logger';
+import { PROVIDER_ERROR_CODES, hasErrorCode } from '../errors/WalletConnectionErrors';
+
+const log = createLogger('SnapService');
+
+// Type for the invokeSnap function from snap-utils
+type InvokeSnapFunction = (params: { method: string; params?: Record<string, unknown> }) => Promise<unknown>;
+
+export interface TransactionOutput {
+  address?: string;
+  value?: string;
+  token?: string;
+  type?: string;
+  data?: string;
+}
+
+export interface SendTransactionParams {
+  network: string;
+  outputs: TransactionOutput[];
+  inputs?: Array<{ txId: string; index: number }>;
+  changeAddress?: string;
+}
+
+export interface Transaction {
+  txId: string;
+  timestamp: number;
+  inputs: Array<{
+    address: string;
+    value: bigint;
+    token: string;
+  }>;
+  outputs: Array<{
+    address: string;
+    value: bigint;
+    token: string;
+  }>;
+  confirmed: boolean;
+}
+
+/**
+ * Custom error class for unauthorized snap errors
+ */
+export class SnapUnauthorizedError extends Error {
+  code: number;
+
+  constructor(message: string, code: number = 4100) {
+    super(message);
+    this.name = 'SnapUnauthorizedError';
+    this.code = code;
+  }
+}
+
+/**
+ * Wraps snap calls with error handling for unauthorized errors
+ */
+async function wrapSnapCall<T>(
+  methodName: string,
+  snapCall: () => Promise<T>
+): Promise<T> {
+  try {
+    return await snapCall();
+  } catch (error) {
+    log.error(`${methodName} failed:`, error);
+
+    // Check for unauthorized error (code 4100)
+    if (hasErrorCode(error, PROVIDER_ERROR_CODES.UNAUTHORIZED)) {
+      throw new SnapUnauthorizedError(
+        'Snap permissions have been revoked or changed. Please reconnect your wallet.',
+        4100
+      );
+    }
+
+    // Re-throw the original error
+    throw error;
+  }
+}
+
+// These will be used by the wallet context with the hooks
+export const WalletServiceMethods = {
+  async getAddress(invokeSnap: InvokeSnapFunction, type: 'index' = 'index', index: number = 0): Promise<string> {
```

Comment:
It won't throw any errors but the type might be `first_empty`, which does not expect index parameter

Reply:
Updated to handle first_empty type, only includes index param when type === 'index'